### PR TITLE
Platform Driver Developement Framework (PDDF)

### PR DESCRIPTION
**Pull request by fk410167**
Signed-off-by: Fuzail Khan <fuzail.khan@broadcom.com>

This change introduces PDDF which is described here,
https://github.com/Azure/SONiC/pull/536

Most of the platform bring up effort goes in developing the platform device drivers, SONiC platform APIs and validating them. Typically each platform vendor writes their own drivers and platform APIs which is very tailor made to that platform. This involves writing code, building, installing it on the target platform devices and testing. Many of the details of the platform are hard coded into these drivers, from the HW spec. They go through this cycle repetitively till everything works fine, and is validated before upstreaming the code.
PDDF aims to make this platform driver and platform APIs development process much simpler by providing a data driven development framework. This is enabled by:

JSON descriptor files for platform data
Generic data-driven drivers for various devices
Generic SONiC platform APIs
Vendor specific extensions for customisation and extensibility

**- Why I did it**
Rapid development and qualification of different platforms on SONiC OS.

**- How I did it**
PDDF generic platform drivers and utils are provided under
sonic-buildimage/platform/pddf/i2c/

PDDF generic user space 2.0 platform APIs are added here,
sonic-buildimage/platform/pddf/platform-api-pddf-base/

Rest of the changes are added to enable the framework.

**- How to verify it**
We will add support for a sample platform to enable verification.

Under PDDF mode, pddf utils can be verified,
# sudo pddf_psuutil <options>
# sudo pddf_fanutil <options>
# sudo pddf_thermalutil <options>
# sudo pddf_ledutil <options>

**- Description for the changelog**

Platform Driver Development Framework (PDDF): Generic kernel device drivers and platform APIs

**- A picture of a cute animal (not mandatory but encouraged)**

Depends on:
https://github.com/Azure/sonic-platform-common/pull/92
https://github.com/Azure/sonic-utilities/pull/940

### CODE REVIEW

File: platform/pddf/platform-api-pddf-base/sonic_platform_pddf_base/pddf_chassis.py [hunk 1]
```diff
@@ -0,0 +1,517 @@
...
+                self.pddf_obj = pddfparse.PddfParse()
+                with open('/usr/share/sonic/platform/pddf/pd-plugin.json') as pd:
+                    self.plugin_data = json.load(pd)
+            except Exception as e:
+                raise Exception("Error: Unable to load PDDF JSON data - %s" % str(e))
+
+        self.platform_inventory = self.pddf_obj.get_platform()
+
+        # Initialize EEPROM
+        self.sys_eeprom = Eeprom(self.pddf_obj, self.plugin_data)
```
Comment (fk410167) [inline platform/pddf/platform-api-pddf-base/sonic_platform_pddf_base/pddf_chassis.py:48]: lgtm[py/call/wrong-number-class-arguments]

File: platform/pddf/platform-api-pddf-base/sonic_platform_pddf_base/pddf_chassis.py [hunk 2]
```diff
@@ -0,0 +1,517 @@
...
+
+        # FANs
+        for i in range(self.platform_inventory['num_fantrays']):
+            for j in range(self.platform_inventory['num_fans_pertray']):
+                fan = Fan(i, j, self.pddf_obj, self.plugin_data)
```
Comment (fk410167) [inline platform/pddf/platform-api-pddf-base/sonic_platform_pddf_base/pddf_chassis.py:53]: lgtm[py/call/wrong-number-class-arguments]

File: platform/pddf/platform-api-pddf-base/sonic_platform_pddf_base/pddf_chassis.py [hunk 3]
```diff
@@ -0,0 +1,517 @@
...
+                self._fan_list.append(fan)
+
+        # PSUs
+        for i in range(self.platform_inventory['num_psus']):
+            psu = Psu(i, self.pddf_obj, self.plugin_data)
```
Comment (fk410167) [inline platform/pddf/platform-api-pddf-base/sonic_platform_pddf_base/pddf_chassis.py:58]: lgtm[py/call/wrong-number-class-arguments]

File: platform/pddf/platform-api-pddf-base/sonic_platform_pddf_base/pddf_chassis.py [hunk 4]
```diff
@@ -0,0 +1,517 @@
...
+            self._psu_list.append(psu)
+
+        # OPTICs
+        for index in range(self.platform_inventory['num_ports']):
+            sfp = Sfp(index, self.pddf_obj, self.plugin_data)
```
Comment (fk410167) [inline platform/pddf/platform-api-pddf-base/sonic_platform_pddf_base/pddf_chassis.py:63]: lgtm[py/call/wrong-number-class-arguments]

File: platform/pddf/platform-api-pddf-base/sonic_platform_pddf_base/pddf_chassis.py [hunk 5]
```diff
@@ -0,0 +1,517 @@
...
+            self._sfp_list.append(sfp)
+
+        # THERMALs
+        for i in range(self.platform_inventory['num_temps']):
+            thermal = Thermal(i, self.pddf_obj, self.plugin_data)
```
Comment (fk410167) [inline platform/pddf/platform-api-pddf-base/sonic_platform_pddf_base/pddf_chassis.py:68]: lgtm[py/call/wrong-number-class-arguments]

File: platform/pddf/platform-api-pddf-base/sonic_platform_pddf_base/pddf_platform.py [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+        self.pddf_data = pddfparse.PddfParse()
+        with open('/usr/share/sonic/platform/pddf/pd-plugin.json') as pd:
+            self.pddf_plugin_data = json.load(pd)
+
+        if not self.pddf_data or not self.pddf_plugin_data:
+            print "Error: PDDF JSON data is not loaded properly ... Exiting"
+            raise ValueError
+
+        PlatformBase.__init__(self)
+        self._chassis = Chassis(self.pddf_data, self.pddf_plugin_data)
```
Comment (fk410167) [inline platform/pddf/platform-api-pddf-base/sonic_platform_pddf_base/pddf_platform.py:37]: lgtm[py/call/wrong-number-class-arguments]

File: platform/pddf/platform-api-pddf-base/sonic_platform_pddf_base/pddf_psu.py [hunk 7]
```diff
@@ -0,0 +1,301 @@
...
+
+        self.pddf_obj = pddf_data
+        self.plugin_data = pddf_plugin_data
+        self.platform = self.pddf_obj.get_platform()
+        self.psu_index = index + 1
+        
+        self._fan_list = []     # _fan_list under PsuBase class is a global variable, hence we need to use _fan_list per class instatiation
+        self.num_psu_fans = int(self.pddf_obj.get_num_psu_fans('PSU{}'.format(index+1)))
+        for psu_fan_idx in range(self.num_psu_fans):
+            psu_fan = Fan(0, psu_fan_idx, pddf_data, pddf_plugin_data, True, self.psu_index)
```
Comment (fk410167) [inline platform/pddf/platform-api-pddf-base/sonic_platform_pddf_base/pddf_psu.py:46]: lgtm[py/call/wrong-number-class-arguments]

**Comment by fk410167** [comment]
Added comment
"lgtm[py/call/wrong-number-class-arguments]"
on various lines to ignore this specific alert as they are not correct.

**Comment by fk410167** [comment]
retest vsimage

**Comment by fk410167** [comment]
retest vsimage please

**Comment by fk410167** [comment]
Need to ignore the 7 LGTM errors.

These errors are flagged because LGTM is trying to match platform classes developed outside of the PDDF framework with the expectations of the framework, which is not really valid.

**Comment by ben-gale** [comment]
@jleveque - can we get this one merged please?

**Comment by ben-gale** [comment]
> @jleveque - can we get this one merged please?

@jleveque - Joe?

**Review by jleveque — CHANGES_REQUESTED** [review]
As comments. Also, please resolve the conflict. Also suggest running all Python files through a tool such as autopep8 (e.g., `autopep8 --max-line-length 120`) to unify code style.

**Comment by fk410167** [comment]
@jleveque - I took care of your comments. Can you review?

**Comment by fk410167** [comment]
retest vsimage please

**Comment by fk410167** [comment]
retest vsimage, vs, broadcom please

**Comment by fk410167** [comment]
retest broadcom please

**Comment by fk410167** [comment]
retest vs please

**Comment by fk410167** [comment]
restest vsimage

**Comment by fk410167** [comment]
restest vsimage please

**Comment by fk410167** [comment]
restest vsimage please

**Comment by fk410167** [comment]
@jleveque 
Can you review this please?

**Comment by fk410167** [comment]
retest vsimage please

**Review by jleveque — APPROVED** [review]

**Pull request merged**